pkg/applicationmanager/controllers/applicationrestore.go

```diff
@@ -674,82 +693,95 @@ func (a *ApplicationRestoreController) restoreVolumes(restore *storkapi.Applicat
+						backupVolInfos, existingRestoreVolInfos, err = a.skipVolumesFromRestoreList(restore, objects, driver, vInfos)
 						if err != nil {
+							log.ApplicationRestoreLog(restore).Errorf("Error while checking pvcs: %v", err)
 							return err
 						}
-						if !skip {
-							objectBasedOnIncludeResources = append(
-								objectBasedOnIncludeResources,
-								o,
-							)
+					}
+					var storageClassesBytes []byte
+					if driverName == "csi" {
+						storageClassesBytes, err = a.downloadObject(backup, backup.Spec.BackupLocation, backup.Namespace, "storageclasses.json", false)
+						if err != nil {
+							log.ApplicationRestoreLog(restore).Errorf("Error in a.downloadObject %v", err)
+							return err
 						}
 					}
-					tempObjects, err := a.getNamespacedObjectsToDelete(
-						restore,
-						objectBasedOnIncludeResources,
-					)
+					preRestoreObjects, err = driver.GetPreRestoreResources(backup, restore, objects, storageClassesBytes)
 					if err != nil {
+						log.ApplicationRestoreLog(restore).Errorf("Error getting PreRestore Resources: %v", err)
 						return err
 					}
-					err = a.resourceCollector.DeleteResources(
-						a.dynamicInterface,
-						tempObjects, updateCr)
-					if err != nil {
-						return err
+
+					// Pre-delete resources for CSI driver
+					if (driverName == "csi" || driverName == "kdmp") && restore.Spec.ReplacePolicy == storkapi.ApplicationRestoreReplacePolicyDelete {
+						objectMap := storkapi.CreateObjectsMap(restore.Spec.IncludeResources)
+						objectBasedOnIncludeResources := make([]runtime.Unstructured, 0)
+						var opts resourcecollector.Options
+						for _, o := range objects {
+							skip, err := a.resourceCollector.PrepareResourceForApply(
+								o,
+								objects,
+								objectMap,
+								restore.Spec.NamespaceMapping,
+								nil, // no need to set storage class mappings at this stage
+								nil,
+								restore.Spec.IncludeOptionalResourceTypes,
+								nil,
+								&opts,
+								restore.Spec.BackupLocation,
+								restore.Namespace,
+							)
+							if err != nil {
+								return err
+							}
+							if !skip {
+								objectBasedOnIncludeResources = append(
+									objectBasedOnIncludeResources,
+									o,
+								)
+							}
+						}
+						tempObjects, err := a.getNamespacedObjectsToDelete(
+							restore,
+							objectBasedOnIncludeResources,
+						)
+						if err != nil {
+							return err
+						}
+						err = a.resourceCollector.DeleteResources(
+							a.dynamicInterface,
+							tempObjects, updateCr)
+						if err != nil {
+							return err
+						}
 					}
-				}
-				// pvc creation is not part of kdmp
-				if driverName != volume.KDMPDriverName {
-					if err := a.applyResources(restore, preRestoreObjects, updateCr); err != nil {
-						return err
+					// pvc creation is not part of kdmp
+					if driverName != volume.KDMPDriverName {
+						if err := a.applyResources(restore, preRestoreObjects, updateCr); err != nil {
```

Comment:
nit: Logically it seems if preRestoreObjects are nil, then we are not doing anything important in this call. But what I see is that, we process a lot for S3+PXD but we are not for NFS+PXD